### PR TITLE
Convert natgateways service to SDKv2

### LINKED ISSUE #3441: Adapt async framework to SDK v2

### LINKED ISSUE #3442: Update NAT gateways service to SDK v2

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Converts the natgateways service to SDKv2 and a new `asyncpoller` framework for long-running operations.

**Which issue(s) this PR fixes**:

Fixes #3442
Fixes #3441

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mboersma** [comment]
I rebased this (temporarily) on #3843 and made the necessary changes to use that apporach for authentication.

**Comment by mboersma** [comment]
This is ready for review if anyone has time.

But I'd like to get #3857 full working as well before we merge this, so we have high confidence in the `asyncpoller` framework.

File: azure/services/asyncpoller/interfaces.go
Comment (mboersma) [reply azure/services/asyncpoller/interfaces.go:38]: It appears we can. I started with a 1:1 port of `async` and refactored away the need for this but must have kept the code.

**Comment by mboersma** [comment]
/hold

For more reviews, testing, and eventual squash.

File: azure/services/natgateways/client.go
Comment (mboersma) [reply azure/services/natgateways/client.go]: Done--set to a constant of 1.

File: azure/services/asyncpoller/asyncpoller.go
Comment (mboersma) [reply azure/services/asyncpoller/asyncpoller.go]: I don't see a reasonable way to get at the response from that context. I removed this function and just use the default time explicitly.

File: azure/services/asyncpoller/asyncpoller_test.go
Comment (mboersma) [reply azure/services/asyncpoller/asyncpoller_test.go]: Added, good call.

File: azure/services/asyncpoller/asyncpoller.go
Comment (mboersma) [reply azure/services/asyncpoller/asyncpoller.go]: Updated: I left a function in place so we could document that using the "retry-after" header would be a better approach.

File: azure/services/asyncpoller/asyncpoller.go
Comment (mboersma) [reply azure/services/asyncpoller/asyncpoller.go]: I left a function in place to document that using the "retry-after" header would be a better approach.

File: azure/services/natgateways/spec.go
Comment (nojnhuh) [inline azure/services/natgateways/spec.go:97]: I suppose we could get rid of this entirely since an empty slice will basically skip over the for loop and return false anyway.

**Review by nojnhuh — COMMENTED** [review]
/lgtm

**Comment by mboersma** [comment]
I squashed the commits and added some additional unit test paths through `CreateOrUpdateAsync`. 

The hope is that the second commit here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/commits/517cfe2773ea7a0aef2ea9b875e832f96ceaf866 will serve as a reference for others to convert services to SDKv2 / `asyncpoller`. Please keep that in mind and suggest improvements, additional comments, etc.

**Comment by nojnhuh** [comment]
LGTM pending review from at least one other person.

**Review by willie-yao — APPROVED** [review]
This looks great! Thanks for your hard work on this @mboersma 
/lgtm

**Comment by mboersma** [comment]
/retest

Looks like a provisioning flake in ci-entrypoint: timeout `# Wait for the kubeconfig to become available.`

**Pull request merged**
